### PR TITLE
[SECURITY-367][JENKINS-21327] Unrestricted Groovy scripting vulnerability

### PR DESCRIPTION
**Pull request by Wadeck**
For the moment this is a **proposal** that need to be discussed / accepted before finalized (no unit test)

- integration of Script Security plugin
  - the sandbox is not usable since the scripts are potentially sent to Node
  - only use the approval queue system
- no change compared to previous version except:
  - core dependency updated from `1.580` to `1.580.1` (for Script Security)
  - instead of requiring Overall/ADMINISTER, now we use Overall/RUN_SCRIPTS
- give possibility to use sub permissions to avoid giving everyong Overall/RUN_SCRIPTS
  - Scriptler/CONFIGURE used to add/edit scripts
  - Scriptler/RUN_SCRIPTS used to run scripts and add/edit Scriptler build steps
  - both new permissions are implied by Overall/RUN_SCRIPTS, so they are "optional"
- enforcement of permission check on different pages
- small modifications / typo correction on the fly (don't hit me 😢)
- alternate approach of [#29](https://github.com/jenkinsci/scriptler-plugin/pull/29) to satisfy the requirements (inspired by it)

[SECURITY-367](https://issues.jenkins-ci.org/browse/SECURITY-367) (private security issue)
[JENKINS-21327](https://issues.jenkins-ci.org/browse/JENKINS-21327)

@reviewbybees @imod @daniel-beck

### CODE REVIEW

File: src/main/webapp/scriptler.css [hunk 1]
```diff
@@ -13,6 +13,7 @@
 .scriptler-link-button {
   background: none;
   border: none;
+  cursor: pointer;
```
Comment (Wadeck) [inline src/main/webapp/scriptler.css]: Concerns the catalog save icon
screenshot_2018-01-11_224301_001

File: src/main/resources/org/jenkinsci/plugins/scriptler/ScriptlerManagement/sidepanel.jelly [hunk 2]
```diff
@@ -31,13 +31,17 @@ THE SOFTWARE.
...
-      <l:task icon="images/24x24/new-package.gif" href="scriptsettings" title="${%Add a new Script}" permission="${app.ADMINISTER}" />
-      <j:if test="${!it.disableRemoteCatalog()}">
-      		<l:task icon="images/24x24/folder.gif" href="catalog" title="${%Remote Script catalogs}" permission="${app.ADMINISTER}" />
+      <j:if test="${it.hasAtLeastOneScriptlerPermission()}">
+        <l:task icon="images/24x24/next.gif" href="${rootURL}/scriptler" title="${%Run/Edit Scripts}" />
       </j:if>
-      <l:task icon="images/24x24/setting.gif" href="settings" title="${%Scriptler settings}" permission="${app.ADMINISTER}" />
-      <l:task icon="plugin/scriptler/images/Git-Icon-1788C.png" href="${rootURL}/scriptler.git" title="${%Git repo}" permission="${app.ADMINISTER}" />
+      <l:hasPermission permission="${it.scriptlerConfigure}">
+        <l:task icon="images/24x24/new-package.gif" href="scriptSettings" title="${%Add a new Script}" />
```
Comment (Wadeck) [inline src/main/resources/org/jenkinsci/plugins/scriptler/ScriptlerManagement/sidepanel.jelly]: The view urls were confusing for the application (two were highlighted), so I use camelCase to avoid this "issue".
screenshot_2018-01-11_224047_001

File: src/main/resources/org/jenkinsci/plugins/scriptler/ScriptlerManagement/runScript.properties [hunk 3]
```diff
@@ -22,7 +22,8 @@
 
 title=Run a script
 intro=Please select a slave you would like to run the script on. You can also modify the script to your needs, the changes will not effect the stored version of it!<br /><br />\
-You can also run a script using GET or POST to the URL /scriptler/run/&lt;your-script-id&gt; with the optional parameters node, script (alternative script text), contentType, and the parameters defined for your stored script.
+You can also run a script using POST to the URL <code>/scriptler/run/&lt;your-script-id&gt;</code> with the optional parameters <code>node</code>, <code>script</code> (alternative script text), <code>contentType</code>, and the parameters defined for your stored script.
```
Comment (Wadeck) [inline src/main/resources/org/jenkinsci/plugins/scriptler/ScriptlerManagement/runScript.properties]: The "{{GET}} or {{POST}}" modification of the documentation is a correction of a modification done in another commit already merged. 

@imod the wiki page could also be changed when the 1.8 will be released in the REST chapter

This was corrected for SECURITY-334 by commit [50072e54c](https://github.com/jenkinsci/scriptler-plugin/commit/50072e54c443b9a894100e8771e492bddaf88a7d)

Comment (imod) [reply src/main/resources/org/jenkinsci/plugins/scriptler/ScriptlerManagement/runScript.properties]: 👍 sorry, I did not find time to take a closer look at this PR yet - will try to find some on the next weekend

Comment (Wadeck) [reply src/main/resources/org/jenkinsci/plugins/scriptler/ScriptlerManagement/runScript.properties]: @imod just a ping in case you forgot this PR

File: src/main/java/org/jenkinsci/plugins/scriptler/util/ScriptHelper.java [hunk 4]
```diff
@@ -79,6 +79,53 @@ public static Script getScript(String id, boolean withSrc) {
...
+     */
+    public static boolean isApproved(String scriptSourceCode){
+        return isApproved(scriptSourceCode, true);
+    }
+    
+    /**
+     * @param scriptSourceCode Source code that must be approved
+     * @param putInApprovalQueueIfNotApprovedYet true means we try to know if the user has permission
+     *                                          to approve the script automatically in case it was not approved yet
+     * @return true iff the script is approved
```
Comment (imod) [inline src/main/java/org/jenkinsci/plugins/scriptler/util/ScriptHelper.java]: typo: `iff` -> `if`

File: src/main/resources/org/jenkinsci/plugins/scriptler/ScriptlerManagement/scriptSettings.properties [hunk 5]
```diff
@@ -26,7 +26,7 @@ intro=\
   Use the ''println'' command to see the output (if you use <tt>System.out</tt>, \
   it will go to the server''s stdout, which is harder to see.). If you think your script might be useful to others too, please share it on <a href="https://github.com/jenkinsci/jenkins-scripts" target="_blank">https://github.com/jenkinsci/jenkins-scripts</a>.
 uploadtext=Select a Groovy script from your local system to be uploaded (*.groovy). 
-Permission = Enable to user with overall permission read
+Permission = Usage in build step
```
Comment (imod) [inline src/main/resources/org/jenkinsci/plugins/scriptler/ScriptlerManagement/scriptSettings.properties]: use same wording as above: `Allow usage in build step`

File: src/main/java/org/jenkinsci/plugins/scriptler/util/ScriptHelper.java (hunk 4 above)
Comment (Wadeck) [reply src/main/java/org/jenkinsci/plugins/scriptler/util/ScriptHelper.java]: `iff` = "if and only if", but I will change it as it's not that clear 👍

**Comment by daniel-beck** [comment]
> btw. has an admin any way to see what scripts are approved already?

Could probably check for any given script, but only a checksum is stored, so no list of approved scripts.

**Comment by Wadeck** [comment]
> btw. has an admin any way to see what scripts are approved already?

Yeah sorry I did not mention that point in the PR. Currently when a user with just Scriptler/RunScripts wants to run a script in the "runScript" page, a warning message is displayed in case the script is not yet approved. (not dynamically updated in case he wants to manually edit the script)

scriptler_not_approved

In case the user has Overall/RunScripts permission, the run of an unapproved script will automatically approve it, so a warning + info message are displayed in such case.

scriptler_will_be_approved

I will try to find a way to have a new icon to display the approval status of script directly in the list.

**Comment by Wadeck** [comment]
In order to display the approval status of the script, I needed to rewrite a bit the icon system used in the list view.
There were also some problems with availability of script that was not really set.

New list:
scriptler_toolbar


With that you have all the useful information at once.

**Comment by Wadeck** [comment]
> I did some testing and one thing I don't like is:

@imod additionally this is corrected after the new commits

**Comment by imod** [comment]
@Wadeck my first attempt to merge this failed because I somehow got lost in the merge conflicts - I'll try to find again time tomorrow...

**Comment by Wadeck** [comment]
@imod normally it's ok now
